File: load-testing/test_helper.rb
# require 'rubygems'
gem "selenium-webdriver"
require 'selenium-webdriver'
require 'rspec'

# include utility functions such as 'page_text', 'try_for', 'fail_safe', ..., etc.
require "#{File.dirname(__FILE__)}/agileway_utils.rb"

# when running in TestWise, it will auto load TestWiseRuntimeSupport, ignore otherwise
if defined?(TestWiseRuntimeSupport)
  ::TestWise::Runtime.load_webdriver_support # for selenium webdriver support
end

# this loads defined page objects under pages folder
require "#{File.dirname(__FILE__)}/pages/abstract_page.rb"
Dir["#{File.dirname(__FILE__)}/pages/*_page.rb"].each { |file| load file }

# The default base URL for running from command line or continuous build process
$BASE_URL = "https://travel.agileway.net"

# This is the helper for your tests, every test file will include all the operation
# defined here.
module TestHelper

  include AgilewayUtils
  if defined?(TestWiseRuntimeSupport)  # TestWise 5+
    include TestWiseRuntimeSupport 
  end

  def browser_type
    if $TESTWISE_BROWSER then
      $TESTWISE_BROWSER.downcase.to_sym
    elsif ENV["BROWSER"]
      ENV["BROWSER"].downcase.to_sym      
    else
      :chrome
    end
  end
  alias the_browser browser_type

  def site_url(default = $BASE_URL)
    $TESTWISE_PROJECT_BASE_URL || ENV["BASE_URL"] || default
  end
	
  def browser_options
    the_browser_type = browser_type.to_s

    if the_browser_type == "chrome"
      the_chrome_options = Selenium::WebDriver::Chrome::Options.new
      # make the same behaviour as Python/JS
      # leave browser open until calls 'driver.quit'
      the_chrome_options.add_option("detach", true)

      if ENV["BROWSER_HEADLESS"] == "true"
        the_chrome_options.add_argument("--headless")
      end

      if defined?(TestWiseRuntimeSupport)
        browser_debugging_port = get_browser_debugging_port() rescue 19218 # default port
        puts("Enabled chrome browser debug port: #{browser_debugging_port}")
        the_chrome_options.add_argument("--remote-debugging-port=#{browser_debugging_port}")
      else
        # puts("Chrome debugging port not enabled.")
      end

      return :options => the_chrome_options

    elsif the_browser_type == "firefox"
      the_firefox_options = Selenium::WebDriver::Firefox::Options.new
      if $TESTWISE_BROWSER_HEADLESS || ENV["BROWSER_HEADLESS"] == "true"
        the_firefox_options.add_argument("--headless")
      end
      return :options => the_firefox_options
    elsif the_browser_type == "ie"
      the_ie_options = Selenium::WebDriver::IE::Options.new
      if $TESTWISE_BROWSER_HEADLESS || ENV["BROWSER_HEADLESS"] == "true"
        # not supported yet?
        # the_ie_options.add_argument('--headless')
      end
      return :options => the_ie_options
    else
      return {}
    end
  end
	
  def driver
    @driver
  end
  
  def browser
    @driver
  end
  
  
  # got to path based on current base url
  def visit(path)
    driver.navigate.to(site_url + path)
  end
  
  def page_text
    driver.find_element(:tag_name => "body").text
  end
  
  def debugging?
    return ENV["RUN_IN_TESTWISE"].to_s == "true" && ENV["TESTWISE_RUNNING_AS"] == "test_case"
  end
  
  # prevent extra long string generated test scripts that blocks execution when running in 
  # TestWise or BuildWise Agent
  def safe_print(str)
    return if str.nil? || str.empty?
    if (str.size < 250)
      puts(str)
      return;
    end
    
    if ENV["RUN_IN_TESTWISE"].to_s == "true" && ENV["RUN_IN_BUILDWISE_AGENT"].to_s == "true"
      puts(str[0..200])
    end
  end
  
  def login_as(user, pass = "testwise")
    driver.find_element(:id, "username").send_keys(user)
    driver.find_element(:id, "password").send_keys(pass)
    driver.find_element(:name, "commit").click
  end

  def logout
    driver.find_element(:link_text, "Sign off").click
  end
  
end

File: playwright-mocha-chrome/test_helper.js
const { test, expect } = require('@playwright/test');
const { chromium } = require('playwright');
var assert = require('assert');

const fs = require('fs');
const path = require('path');

function debuggingPort() {
    var port_number = Math.floor((Math.random() * 40000) + 16000);
    var port_number_str = "" + port_number;
    fs.writeFile(portFile(), port_number_str, function(err) {});
    return port_number_str;
}

function getDebuggingPort() {
  return fs.readFileSync(portFile());
}

function portFile() {
  var port_number_file_path = path.join(__dirname, "port_number.txt")
  return port_number_file_path;
}

module.exports = {

    browserType: function() {
        if (process.env.BROWSER) {
            return process.env.BROWSER;
        } else {
            return "chrome";
        }
    },

    chromeOptions: function() {
        opts = new chrome.Options();
        opts.addArguments("--remote-debugging-port=" + debuggingPort());
        return opts;
    },

    debuggingChromeOptions: function() {
        opts = new chrome.Options();
        var debugging_address = "127.0.0.1:" + getDebuggingPort();
        opts.options_["debuggerAddress"] = debugging_address;
        return opts;
    },

    is_debugging: function() {
        return (process.env.RUN_IN_TESTWISE == "true" && process.env.TESTWISE_RUNNING_AS == "test_case");
    },

    site_url: function() {
        if (process.env.BASE_URL) {
            return process.env.BASE_URL;
        } else {
            return "https://travel.agileway.net";
        }
    },

};

// BEGIN: user functions
async function login(driver, username, password) {
    await driver.findElement(webdriver.By.name('username')).sendKeys(username);
    await driver.findElement(webdriver.By.name('password')).sendKeys(password);
    await driver.findElement(webdriver.By.name('commit')).click();
}

// END: user functions


// BEGIN: module exports
module.exports.login = login;

// END: module exports